textile/features.textile

```diff
@@ -1499,6 +1500,29 @@ h4. ChannelMetrics
 ** @(CHM2e)@ @publishers@ integer - the number of realtime attachments which are able to publish messages to the channel (that is, they have the `publish` capability and have not specified a @ChannelMode@ that excludes @PUBLISH@)
 ** @(CHM2f)@ @subscribers@ integer - the number of realtime attachments which are receiving messages on the channel (that is, they have the `subscribe` capability and have not specified a @ChannelMode@ that excludes @SUBSCRIBE@)
 
+h4. BatchResult
+
+* @(BAR1)@ Contains information about the results of a batch operation
+** @(BAR2)@ The attributes of @BatchResult@ consist of:
+** @(BAR2a)@ @successCount@ number - the number of successful operations
+** @(BAR2b)@ @failureCount@ number - the number of unsuccessful operations
+** @(BAR2c)@ @results@ array - an array of results for the batch operation
+
+h4. BatchSpec
+
+* @(BSP1)@ Describes the messages that should be published by a batch publish operation, and the channels to which they should be published
+* @(BSP2)@ The attributes of @BatchSpec@ consist of:
+** @(BSP2a)@ @channels@ an array of strings - the names of the channels to which all of the messages contained in the @messages@ attribute should be published
+** @(BSP2b)@ @messages@ an array of @Message@ objects - the names of the channels to which all of the messages contained in the @messages@ attribute should be published
+
+h4. BatchPublishResult
+
+* @(BPR1)@ Contains information about the result of publishes requested by a single @BatchSpec@ for a single channel
```

Comment:
Yeah you're right - it's not supposed to be for a single channel, that was my bad. It is indeed possible to correlate messages and channels since the returned `BatchPublishResult` array is in the same order as the provided `BatchSpec` array

Reply:
I remain a bit confused about the structure of the result of `batchPublish(BatchPublishSpec)`.

Would you mind confirming whether this is correct?

> that is, given a `BatchSpec` for publishing $m$ messages to $n$ channels, the `BatchResult.results` array will contain $mn$ `BatchPublishResult` objects

If so, I'm not sure I understand this comment of yours:

> the returned `BatchPublishResult` array is in the same order as the provided `BatchSpec` array

`BatchSpec` is a 2-dimensional array but `BatchPublishResult` is 1-dimensional. So to say that they have the "same order" doesn’t make sense to me — there must be some sort of flattening going on?

Reply:
> that is, given a BatchSpec for publishing _m_ messages to _n_ channels, the BatchResult.results array will contain _mn_ BatchPublishResult objects

(right now it will have _n_ since all messages for one channel are bundled together and sent atomically, cf reply to the messageId thread)

Reply:
> BatchPublishSpec is a 2-dimensional structure (has messages and channels) but BatchResult.results is 1-dimensional. So to say that they have the "same order" doesn’t automatically make sense to me — there must be some sort of flattening going on?

What I meant was: If you post an array of `BatchSpecs`, you will get an array of `BatchResult`s in the same order as the corresponding array of `BatchSpec`s. For each individual `BatchSpec` you can match on channel name within the `BatchResult.results` array to determine which publishes were successful/unsuccessful. The code example [here](https://github.com/ably/docs/pull/1450#issuecomment-1161500432) may be helpful.